src/core/layertree/qgslayertreemodel.cpp

```diff
@@ -1299,10 +1303,41 @@ QList<QgsLayerTreeModelLegendNode*> QgsLayerTreeModel::layerLegendNodes( QgsLaye
 
 void QgsLayerTreeModel::legendInvalidateMapBasedData()
 {
+  // we have varying icon sizes, and we want icon to be centered and
+  // text to be left aligned, so we have to compute the max width of icons
+  //
+  // we do that for nodes who share a common parent
+  //
+  // we do that here because for symbols with size defined in map units
+  // the symbol sizes changes depends on the zoom level
+
+  QList<QgsSymbolV2LegendNode*> symbolNodes;
+  QMap<QString, int> widthMax;
   foreach ( const LayerLegendData& data, mLegend )
   {
     foreach ( QgsLayerTreeModelLegendNode* legendNode, data.originalNodes )
+    {
       legendNode->invalidateMapBasedData();
+      QgsSymbolV2LegendNode* n = dynamic_cast<QgsSymbolV2LegendNode*>( legendNode );
+      if ( n )
+      {
+        n->setParent( this ); // map scale are in the model, so the parent needs to be set
```

Comment:
No it's not necessary, you're right. 

I added this because when the symbol size computation was done in QgsDefaultVectorLayerLegend::createLayerTreeModelLegendNodes the parent was not set... and I wanted to make sure :)